### PR TITLE
e2e: Runner is able to start and stop Tendermint testnet processes on non-local machines

### PR DESCRIPTION
**Pull request by williambanfield**
This pull request adds functionality to the e2e runner process to start and stop tendermint nodes on non-local machines.

To achieve this goal, it makes the following set of changes:

* Adds a set of methods for starting and stopping Tendermint to the `infra.Provider` interface and implements them by both the `docker.Provider` and `digitalocean.Provider`. The digitalocean provider expects Tendermint to be running under systemd and is started/stopped/killed using interactions with systemd.
* Adds a set of ssh functionality to the e2e test runner to allow interaction with the remove Tendermint nodes. This ssh functionality expects an `ssh-agent` to be running on the host machine and uses keys from the ssh-agent to connect to the remote. This allows operators of the e2e runner to either run the process locally or to run it on a remote orchestrator machine and grant it permission to ssh to additional nodes using ssh-agent forwarding.
* moves the `dockerComposeExec` functionality to the `docker` package. The remaining uses of this new `docker.ComposeExec` function are all targets for replacement with provider methods.
* Updates the e2e testing to pass the `InfrastructureData` and `InfrastructureType` to the test subprocess.

Closes: #9790 

---

#### PR checklist

- [ ] Tests written/updated, or no tests needed
- [ ] `CHANGELOG_PENDING.md` updated, or no changelog entry needed
- [ ] Updated relevant documentation (`docs/`) and code comments, or no
      documentation updates needed

### CODE REVIEW

File: test/e2e/pkg/ssh/ssh.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+	}
+	am := make([]ssh.AuthMethod, 0, len(signers))
+	for _, signer := range signers {
+		am = append(am, ssh.PublicKeys(signer))
+	}
+	return &ssh.ClientConfig{
+		User:              "root",
+		HostKeyCallback:   hkcWrapper(hkc),
+		Auth:              am,
+		HostKeyAlgorithms: []string{ssh.KeyAlgoED25519},
```
Comment (thanethomson) [inline test/e2e/pkg/ssh/ssh.go:60]: Does this mean that keys always have to be ed25519?

File: test/e2e/pkg/testnet.go [hunk 2]
```diff
@@ -460,7 +459,7 @@ func (n Node) AddressRPC() string {
 
 // Client returns an RPC client for a node.
 func (n Node) Client() (*rpchttp.HTTP, error) {
-	return rpchttp.New(fmt.Sprintf("http://127.0.0.1:%v", n.ProxyPort), "/websocket")
+	return rpchttp.New(fmt.Sprintf("http://%s:%v", n.IP, n.ProxyPort), "/websocket")
```
Comment (thanethomson) [inline test/e2e/pkg/testnet.go:462]: IIRC this change caused problems with the Docker-based backend previously on some people's machines. @jmalicevic do you perhaps remember whether you managed to resolve those problems?

File: test/e2e/tests/e2e_test.go [hunk 3]
```diff
@@ -73,15 +73,32 @@ func loadTestnet(t *testing.T) e2e.Testnet {
...
 	}
-
+	ifdType := os.Getenv("INFRASTRUCTURE_DATA")
+	ifdFile := os.Getenv("INFRASTRUCTURE_FILE")
+	if ifdFile == "" && ifdType != "docker" {
+		t.Skip("INFRASTRUCTURE_DATA not set and INFRASTRUCTURE_TYPE is not docker")
+	}
+	if !filepath.IsAbs(ifdFile) {
+		manifestFile = filepath.Join("..", manifestFile)
+	}
```
Comment (thanethomson) [inline test/e2e/tests/e2e_test.go:83]: Just so I understand how this works, why do we need to look the file up relative to the parent directory?

File: test/e2e/pkg/infra/digitalocean/digitalocean.go [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+)
+
+const (
+	sshPort     = 22
+	testappName = "testappd"
+)
+
+var _ infra.Provider = &Provider{}
+
+// Provider implements a docker-compose backed infrastructure provider.
```
Comment (thanethomson) [inline test/e2e/pkg/infra/digitalocean/digitalocean.go:20]:
```suggestion
// Provider implements a DigitalOcean-backed infrastructure provider.
```

File: test/e2e/pkg/infra/digitalocean/digitalocean.go [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+func (p *Provider) Setup() error {
+	return nil
+}
+
+// Noop currently. Node creation is currently performed externally to the e2e test tool.
+func (p Provider) CreateNode(ctx context.Context, n *e2e.Node) error {
+	return nil
+}
+func (p Provider) StartTendermint(ctx context.Context, n *e2e.Node) error {
+	return e2essh.Exec(p.SSHConfig, fmt.Sprintf("%s:%d", n.IP, sshPort), fmt.Sprintf("systemctl start %s", testappName))
```
Comment (thanethomson) [inline test/e2e/pkg/infra/digitalocean/digitalocean.go:37]: How long does it roughly take to initiate the SSH connection here?

We can test out this SSH-based approach, but if the process manipulation operations take too long to execute they may interfere with the tests in unexpected ways.

If the SSH-based approach works too slowly we should consider adding some kind of proxy service to the nodes that can be accessed via HTTP and that has the appropriate privileges to restart the test app.

File: test/e2e/tests/e2e_test.go (hunk 3 above)
Comment (williambanfield) [reply test/e2e/tests/e2e_test.go]: Ah, this was cribbed from the docker-compose logic but doesn't make sense here. We should just require that flag to be set to an absolute path.

File: test/e2e/pkg/ssh/ssh.go (hunk 1 above)
Comment (williambanfield) [reply test/e2e/pkg/ssh/ssh.go]: Yes, I intended to remove this but appear to have missed it. Removed. When not set it uses a sensible default.

File: test/e2e/pkg/testnet.go (hunk 2 above)
Comment (williambanfield) [reply test/e2e/pkg/testnet.go:462]: I'm seeing that as well. Taking a look.

File: test/e2e/pkg/infra/digitalocean/digitalocean.go (hunk 5 above)
Comment (williambanfield) [reply test/e2e/pkg/infra/digitalocean/digitalocean.go]: The docker version of this uses port mapping with different ports for each process and all ports are exposed on localhost. I've added a new field `ExternalIP` that stores the reachable IP for the node and replaced the old `IP` field with `InternalIP` to specify the IP address internal to the network. It is difficult to harmonize the docker and DO version of this without running the process in the same network space as the docker-compose network space which would be a bit more complex change.

File: test/e2e/pkg/testnet.go (hunk 2 above)
Comment (williambanfield) [reply test/e2e/pkg/testnet.go]: Fixed.

**Pull request closed without merging**